File: app/src/main/java/com/oliver/weatherapp/data/repositories/WeatherRepository.java
package com.oliver.weatherapp.data.repositories;

import android.arch.lifecycle.LiveData;
import android.util.Log;

import com.oliver.weatherapp.AppExecutors;
import com.oliver.weatherapp.data.local.dao.WeatherDao;
import com.oliver.weatherapp.data.local.model.WeatherEntry;
import com.oliver.weatherapp.data.remote.WeatherDataSource;

import java.util.Calendar;
import java.util.Date;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

public class WeatherRepository {

    private static final String TAG = WeatherRepository.class.getSimpleName();
    private final AppExecutors mExecutors;
    private final WeatherDao mWeatherDao;
    // For Singleton instantiation
    private static final Object LOCK = new Object();
    private static WeatherRepository sInstance;
    private final WeatherDataSource mWeatherDataSource;
    private Set<Long> mInitializedForCity = new HashSet<>();

    public static WeatherRepository getInstance(AppExecutors executors,
                                                WeatherDao dao,
                                                WeatherDataSource weatherDataSource) {
        Log.d(TAG, "Get the WeatherRepository");
        if (sInstance == null) {
            synchronized (LOCK) {
                sInstance = new WeatherRepository(executors, dao, weatherDataSource);
                Log.d(TAG, "Made new WeatherRepository");
            }
        }
        return sInstance;
    }

    private WeatherRepository(AppExecutors executors,
                              WeatherDao dao,
                              WeatherDataSource weatherDataSource) {
        mExecutors = executors;
        mWeatherDao = dao;
        mWeatherDataSource = weatherDataSource;
        observeNewWeatherData();
    }

    private void observeNewWeatherData() {
        mWeatherDataSource.getWeather().observeForever(weatherEntries ->
                mExecutors.diskIO().execute(() -> {
                    Log.d(TAG, "observeNewWeatherData: " + (weatherEntries != null ? weatherEntries.length : 0));
                    deleteOldData();

                    mWeatherDao.bulkInsert(weatherEntries);
                }));

    }

    private void deleteOldData() {
        Date today = new Date();
        Log.d(TAG, "deleteOldData: today: " + today);
        mWeatherDao.deleteOldWeather(today);
    }

    public LiveData<List<WeatherEntry>> getForecast(long cityID, double latitude, double longitude) {

        initialize(cityID, latitude, longitude);
//        mWeatherDataSource.fetchForecast(cityID, latitude, longitude);
        return mWeatherDao.getForecast(cityID);
    }


    private void initialize(long cityID, double latitude, double longitude) {
        // initialize only once for instance and for city
        if (mInitializedForCity.contains(cityID)) return;
        mInitializedForCity.add(cityID);

        mExecutors.diskIO().execute(() -> {
            if (isFetchNeeded(cityID)) {
                mWeatherDataSource.fetchForecast(cityID, latitude, longitude);
            }
        });
    }

    /**
     * Need to update weather information if we don't have weather data for today + 5 days
     * if we have at least one record - don't update the weather information
     * @param cityID - id for the city to check cached forecast
     * @return true if we need to perform data update, false otherwise
     */
    private boolean isFetchNeeded(long cityID) {
        Calendar requiredDate = Calendar.getInstance();

        requiredDate.add(Calendar.DAY_OF_MONTH, 5);
        requiredDate.set(Calendar.HOUR_OF_DAY, 0);
        requiredDate.set(Calendar.MINUTE, 0);
        requiredDate.set(Calendar.SECOND, 0);
        requiredDate.set(Calendar.MILLISECOND, 0);

        int count = mWeatherDao.countFutureWeatherForCity(cityID, requiredDate.getTime());
        boolean isFetchNeeded = count == 0;
        Log.d(TAG, "isFetchNeeded: " + isFetchNeeded +
                " requiredUpdateDate: " + requiredDate.getTime() +
                " recordsCount: " + count);
        return isFetchNeeded;
    }
}

File: app/src/main/java/com/oliver/weatherapp/data/repositories/WeatherRepository.kt
package com.oliver.weatherapp.data.repositories

import android.arch.lifecycle.LiveData
import com.oliver.weatherapp.AppExecutors
import com.oliver.weatherapp.data.local.dao.WeatherDao
import com.oliver.weatherapp.data.local.model.WeatherEntry
import com.oliver.weatherapp.data.remote.WeatherDataSource
import timber.log.Timber
import java.util.*

class WeatherRepository (
        private val executors: AppExecutors,
        private val weatherDao: WeatherDao,
        private val weatherDataSource: WeatherDataSource)
{
    private val isInitializedForCity = HashSet<Long>()

    init {
        observeNewWeatherData()
    }

    private fun observeNewWeatherData() {
        weatherDataSource.getWeather().observeForever { weatherEntries ->
            executors.diskIO().execute {
                Timber.d("observeNewWeatherData: ${weatherEntries?.size }")
                deleteOldData()

                weatherEntries?.let {
                    weatherDao.bulkInsert(*it)
                }

            }
        }
    }

    private fun deleteOldData() {
        val today = Date()
        Timber.d("deleteOldData: today: $today")
        weatherDao.deleteOldWeather(today)
    }

    fun getForecast(cityID: Long, latitude: Double, longitude: Double): LiveData<List<WeatherEntry>> {

        initialize(cityID, latitude, longitude)
        //        weatherDataSource.fetchForecast(cityID, latitude, longitude);
        return weatherDao.getForecast(cityID)
    }


    private fun initialize(cityID: Long, latitude: Double, longitude: Double) {
        // initialize only once for instance and for city
        if (isInitializedForCity.contains(cityID)) return
        isInitializedForCity.add(cityID)

        executors.diskIO().execute {
            if (isFetchNeeded(cityID)) {
                weatherDataSource.fetchForecast(cityID, latitude, longitude)
            }
        }
    }

    /**
     * Need to update weather information if we don't have weather data for today + 5 days
     * if we have at least one record - don't update the weather information
     * @param cityID - id for the city to check cached forecast
     * @return true if we need to perform data update, false otherwise
     */
    private fun isFetchNeeded(cityID: Long): Boolean {
        val requiredDate = Calendar.getInstance()

        requiredDate.add(Calendar.DAY_OF_MONTH, 5)
        requiredDate.set(Calendar.HOUR_OF_DAY, 0)
        requiredDate.set(Calendar.MINUTE, 0)
        requiredDate.set(Calendar.SECOND, 0)
        requiredDate.set(Calendar.MILLISECOND, 0)

        val count = weatherDao.countFutureWeatherForCity(cityID, requiredDate.time)
        val isFetchNeeded = count == 0
        Timber.d("isFetchNeeded: $isFetchNeeded " +
                "requiredUpdateDate: ${requiredDate.time} " +
                "recordsCount: $count")
        return isFetchNeeded
    }

    fun forceWeatherRefresh(cityID: Long, latitude: Double, longitude: Double) {
        executors.diskIO().execute {
            weatherDao.deleteWeatherForCity(cityID)
            weatherDataSource.fetchForecast(cityID, latitude, longitude)
        }
    }
}
